src/response.rs

```diff
@@ -291,27 +336,37 @@ impl Response {
             self.length
         };
 
-        let stream = self.stream.into_inner().unwrap();
-        let unit = self.unit;
+        let unit = &self.unit;
         let result = stream.set_read_timeout(unit.agent.config.timeout_read);
         if let Err(e) = result {
             return Box::new(ErrorReader(e)) as Box<dyn Read + Send>;
         }
         let deadline = unit.deadline;
-        let stream = DeadlineStream::new(*stream, deadline);
 
-        let body_reader: Box<dyn Read + Send> = match (use_chunked, limit_bytes) {
-            (true, _) => Box::new(PoolReturnRead::new(
-                &unit.agent,
-                &unit.url,
-                ChunkDecoder::new(stream),
-            )),
-            (false, Some(len)) => Box::new(PoolReturnRead::new(
-                &unit.agent,
-                &unit.url,
-                LimitedRead::new(stream, len),
-            )),
-            (false, None) => Box::new(stream),
+        let body_reader: Box<dyn Read + Send> = match stream {
+            ResponseStream::Stream(stream) => {
+                let stream = DeadlineStream::new(*stream, deadline);
+
+                match (use_chunked, limit_bytes) {
+                    (true, _) => Box::new(PoolReturnRead::new(
+                        &unit.agent,
+                        &unit.url,
+                        ChunkDecoder::new(stream),
+                    )),
+                    (false, Some(len)) => Box::new(PoolReturnRead::new(
+                        &unit.agent,
+                        &unit.url,
+                        LimitedRead::new(stream, len),
+                    )),
+                    (false, None) => Box::new(stream),
+                }
+            }
+            ResponseStream::Buffered(buffer) => match (use_chunked, limit_bytes) {
+                (true, _) => Box::new(ChunkDecoder::new(buffer)),
+                (false, Some(len)) => Box::new(LimitedRead::new(buffer, len)),
+                (false, None) => Box::new(buffer),
```

Comment:
For the Buffered case, `(use_chunked, limit_bytes)` should always be `false, Some(n)`. If were using chunked encoding, that means we don't know the length up front, so we can't say "we know we've already read the full body" based solely on the headers. Note: this means the `let length = get_header(&headers, "content-length")` calculation below on line 553 has to be more refined to use the same logic as here. Specifically: we use content-length only if `Transfer-Encoding` is not set ("[If a message is received with both a Transfer-Encoding and a Content-Length header field, the Transfer-Encoding overrides the Content-Length](https://httpwg.org/specs/rfc9112.html#message.body.length)")

By the same logic, limit_bytes should always be `Some` because if we're in the Buffered case, we know that `Content-Length` is present.

Reply:
Yes. I'll fix this.